### PR TITLE
feat(website): add stretchable layout for playground

### LINKED ISSUE #1846: [FEATURE]  Add Stretchable Sidebar to Playground
### Why do we need this improvement?

We need this improvement to enhance the user experience and usability of the Playground section in our project. Currently, users have the option to hide the entire sidebar, but they cannot resize or stretch the sidebar to view the full code content. This limitation can be frustrating for users, especially when they need to view longer sections of code without excessive scrolling.

By adding a stretchable sidebar feature, users would have the flexibility to expand the width of the sidebar horizontally, allowing them to view the entire code content comfortably. This enhancement would improve the overall usability of the Playground section and make it more intuitive and user-friendly.

this improvement is necessary to provide a smoother and more efficient experience for users interacting with the Playground section, ultimately enhancing their ability to navigate and interact with the code effectively.

### How will this change help?

This change will help improve the usability and functionality of the Playground section in our project in several ways:

1. Enhanced Viewing Experience: The addition of a stretchable sidebar will allow users to expand the width of the sidebar horizontally, enabling them to view the entire code content without having to scroll excessively. This will enhance the overall viewing experience and make it easier for users to navigate and understand the code.

2. Increased Efficiency: With the ability to stretch the sidebar, users can quickly access and review longer sections of code without the need for repetitive scrolling. This will save time and effort for users, increasing their efficiency when interacting with the Playground section.

3. Improved Accessibility: The stretchable sidebar feature will make the code content more accessible to users with varying screen sizes and resolutions. Users will be able to adjust the width of the sidebar according to their preferences and viewing devices, ensuring optimal readability and usability for all users.

4. Better Code Analysis: By providing users with a clear and unobstructed view of the entire code content, the stretchable sidebar will facilitate better code analysis and comprehension. Users will be able to focus on understanding the code logic and structure without distractions or limitations imposed by the interface.

this change will significantly enhance the usability, efficiency, and accessibility of the Playground section, ultimately improving the overall user experience and satisfaction with our project.

### Screenshots

image
image

### How could it be implemented/designed?

1. **User Interface Design:**
   - Add a draggable handle or separator between the sidebar and the main content area.
   - When users hover over the handle, display a cursor indicating that the handle is draggable.
   - Implement visual feedback, such as a change in cursor appearance or highlighting, to indicate the dragging action.

2. **Resize Functionality:**
   - Allow users to click and drag the handle horizontally to resize the width of the sidebar.
   - Implement logic to dynamically adjust the width of the sidebar and the content area as the handle is dragged.

3. **Constraints and Limits:**
   - Set minimum and maximum widths for the sidebar to prevent it from becoming too narrow or too wide, ensuring a usable and aesthetically pleasing layout.
   - Consider responsive design principles to accommodate different screen sizes and resolutions.

4. **Persistence:**
   - Save the user's preferred sidebar width in local storage or as part of their user settings, so that it persists across sessions.
   - Ensure that the selected sidebar width is maintained when navigating between pages or refreshing the browser.

5. **Accessibility Considerations:**
   - Ensure that the stretchable sidebar feature is accessible to users with disabilities by providing keyboard navigation options and ensuring compatibility with screen readers.
   - Implement appropriate aria roles and attributes to convey the purpose and state of the stretchable sidebar to assistive technologies.

6. **Testing and Iteration:**
   - Conduct usability testing with a diverse group of users to gather feedback on the implementation and design of the stretchable sidebar feature.
   - Iterate on the design based on user feedback and make necessary adjustments to improve usability and user satisfaction.

the stretchable sidebar feature can be implemented and designed to provide users with a flexible and intuitive way to resize the sidebar according to their preferences, enhancing the overall user experience of the Playground section in our project.

### 🚧 Breaking changes

Yes

### 👀 Have you checked for similar open issues?

- [X] I checked and didn't find a similar issue

### 🏢 Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/asyncapi/.github/blob/master/CONTRIBUTING.md)

### Are you willing to work on this issue?

No, someone else can work on it.

### PR DESCRIPTION
**Pull request by jerensl**
## Overview
In the context of stretchable editor space playground, I decided to use framer-motion because they really care about [accesibility](https://www.framer.com/motion/guide-accessibility/) and [performance](https://www.framer.com/motion/guide-reduce-bundle-size/), why I really like working this library is how they're just giving us some utility function which can be expanded for broader use cases, I love that because it's enabled us as a developer to be more creative when dealing with animation and interaction.

For the stretchable alone I'm using [useTransform](https://www.framer.com/motion/use-transform/), it's seems counterintuitive in the first place but after you understand it's just a pub sub-model, it's a bit clever how they were implementing it that way.

In the context of state management, I am using useReducer purely because of familiarity with the built-in state management by React, I am also using [Immer](https://immerjs.github.io/immer/example-setstate) for maintaining an immutability state which is an good pattern in react when the state must be immutable and not directly changed, it also enables us to work in a complex nested state, I'm also using typescript to prevent any mutation outside of useReducer

## Functional Requirements
1. User Interface Design:
- [x] Add a draggable handle or separator between the editor space.
- [x] When users hover over the handle, display a cursor indicating that the handle is draggable.
- [x] Implement visual feedback, such as a change in cursor appearance or highlighting, to indicate the dragging action.
2. Resize Functionality:
- [x] Allow users to click and drag the handle horizontally to resize the width of the editor space.
- [x] Implement logic to dynamically adjust the width of the editor space and the content area as the handle is dragged.
3. Constraints and Limits:
- [x] Set minimum and maximum widths for the sidebar to prevent it from becoming too narrow or too wide, ensuring a usable and aesthetically pleasing layout.
- [x] Consider responsive design principles to accommodate different screen sizes and resolutions.
4. Persistence:
- [x] Save the user's preferred editor space width in local storage or as part of their user settings, so that it persists across sessions.
- [x] Ensure that the selected editor space width is maintained when navigating between pages or refreshing the browser.
5. Accessibility Considerations:
- [ ] Ensure that the stretchable editor space feature is accessible to users with disabilities by providing keyboard navigation options and ensuring compatibility with screen readers.
- [x] Implement appropriate aria roles and attributes to convey the purpose and state of the stretchable editor space to assistive technologies.

## Non-Functional Requirements
- [x] Add framer-motion for animation and interaction
- [x] Add Stretchable Layout using framer-motion.
- [x] Integration with Playground.
- [x] Add bundler analyzer to detect whether the library added is implementing tree shaking or not.
- [x] Add measurement on the width of the container, looks like [this library](https://usehooks.com/usemeasure) implemented tree shaking 
- [x] Refactor the HTML element to use a semantic element or [role-based](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles) if not possible.
- [x] Add state management with Immutability State using Immer for managing complex states.
- [x] Adding an indicator that indicates the editor codespace/options is open
- [x] Add a global state for the device screen type to maintain responsiveness.
- [x] Find the solution for storing resizable to global state with percentage not pixel units

## Related Issue
Resolved #1846

### CODE REVIEW

**Comment by jerensl** [comment]
@devilkiller-ag To accommodate max/min width I need to adjust the position of the layout and put output options on the left side of the panel, also I am not sure about the number for now I put 20% minimum appearance for either input/output editor, did you have an opinion about this one?

**Comment by jerensl** [comment]
@devilkiller-ag can I get a review or feedback at least? it's been a while

**Comment by devilkiller-ag** [comment]
> @devilkiller-ag can I get a review or feedback at least? it's been a while

Hi @jerensl Apologies for the late review. I got stuck with some work. I will do the review by tomorrow.

**Comment by jerensl** [comment]
> > @devilkiller-ag can I get a review or feedback at least? it's been a while
> 
> Hi @jerensl Apologies for the late review. I got stuck with some work. I will do the review by tomorrow.

Take your time, I did it just for visibility because sometimes also notifications will not appear on me, and lose track of some PR. No need to hurry :)

**Comment by devilkiller-ag** [comment]
> @devilkiller-ag To accommodate max/min width I need to adjust the position of the layout and put output options on the left side of the panel, also I am not sure about the number for now I put 20% minimum appearance for either input/output editor, did you have an opinion about this one?

We can put a 30% minimum appearance because the input editor generally will have a deeply nested code, and its formatting keeps most code lines after 3-4 tab spaces. Can you try it if it looks good or not?

File: modelina-website/src/store/useLayoutStore.tsx [hunk 1]
```diff
@@ -0,0 +1,153 @@
+import type { Draft, Immutable } from 'immer';
```
Comment (devilkiller-ag) [inline modelina-website/src/store/useLayoutStore.tsx:1]: Why do we need this?

**Review by devilkiller-ag — COMMENTED** [review]
The code changes look good to me. Can you change the minimum appearance for the input/output editor to 30% and check? Also, why do we need to use `immer`?

**Comment by devilkiller-ag** [comment]
> 9\. Accessibility Considerations:
> 
> * [ ]  Ensure that the stretchable editor space feature is accessible to users with disabilities by providing keyboard navigation options and ensuring compatibility with screen readers.

Is it feasible to add this accessibility feature?

Again sorry for the late review, I was stuck in some work the past two weeks.

File: modelina-website/src/store/useLayoutStore.tsx (hunk 1 above)
Comment (jerensl) [reply modelina-website/src/store/useLayoutStore.tsx:1]: It's one of react design's core principles which is [immutability in state](https://react.dev/learn/updating-objects-in-state), sometimes when the state is not immutable it also doesn't trigger the re-rendering process, the state is updated but not changed on the UI

Comment (jerensl) [reply modelina-website/src/store/useLayoutStore.tsx:1]: Also, object reference uses heap allocation when the object property value changes, it's still using the same memory address

**Comment by jerensl** [comment]
> > 9. Accessibility Considerations:
> > 
> > * [ ]   Ensure that the stretchable editor space feature is accessible to users with disabilities by providing keyboard navigation options and ensuring compatibility with screen readers.
> 
> Is it feasible to add this accessibility feature?

I'm not sure either, but I'm already implementing some semantic and basic aria roles as well. Just not be sure how much it helps and how to measure it precisely

> Again sorry for the late review, I was stuck in some work the past two weeks.

No worries

**Comment by devilkiller-ag** [comment]
image

In the mobile view, When accessing the `output navigation` (last icon) from `editor options`, it is always hidden by the opened-up editor options. So, for mobile view, I suggest either hiding its icon when the editor option is open or switching it to the output screen with the `output option` opened up on clicking its icon. Rest everything seems perfect to me.

@jonaslagoni what's your view on this?

**Review by devilkiller-ag — CHANGES_REQUESTED** [review]
image

In the mobile view, When accessing the `output navigation` (last icon) from `editor options`, it is always hidden by the opened-up editor options. So, for mobile view, I suggest either hiding its icon when the editor option is open or switching it to the output screen with the `output option` opened up on clicking its icon. Rest everything seems perfect to me.

@jonaslagoni what's your view on this?

**Review by jonaslagoni — COMMENTED** [review]
>In the mobile view, When accessing the output navigation (last icon) from editor options, it is always hidden by the opened-up editor options. So, for mobile view, I suggest either hiding its icon when the editor option is open or switching it to the output screen with the output option opened up on clicking its icon. Rest everything seems perfect to me.
> 
>@jonaslagoni what's your view on this? 

I agree, never have a button unclickable :v:

Other then that this looks good .v:

**Comment by jonaslagoni** [comment]
@jerensl seem like all this needs is removing the unclickable buttons 🤔

**Comment by jerensl** [comment]
@devilkiller-ag @devilkiller-ag, I apologize for the late response, I already made changes for mobile and hid the output options when the user clicked on the general options, what do you think?

**Review by jonaslagoni — COMMENTED** [review]
Looks good to me :)

**Comment by jonaslagoni** [comment]
@devilkiller-ag anything from you?

**Comment by devilkiller-ag** [comment]
@jonaslagoni All good from my side too!! I was just waiting for the deployment preview to work.  Thanks, @jerensl for working on this issue ✨

**Review by jonaslagoni — APPROVED** [review]

**Pull request merged**